flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/SubstringIndexFunction.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.data.binary.BinaryStringData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import org.apache.commons.lang3.StringUtils;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#SUBSTRING_INDEX}. */
+@Internal
+public class SubstringIndexFunction extends BuiltInScalarFunction {
+    public SubstringIndexFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.SUBSTRING_INDEX, context);
+    }
+
+    public @Nullable StringData eval(
+            @Nullable StringData expr, @Nullable StringData delim, @Nullable Integer count) {
+        if (expr == null || delim == null || count == null) {
+            return null;
+        }
+
+        if (count == 0) {
+            return BinaryStringData.EMPTY_UTF8;
+        }
+        String str = expr.toString();
+        if (str.isEmpty()) {
+            return BinaryStringData.EMPTY_UTF8;
+        }
+        String delimiter = delim.toString();
+        if (delimiter.isEmpty()) {
+            return BinaryStringData.EMPTY_UTF8;
+        }
+
+        try {
+            if (count > 0) {
+                // get index of the count-th occurrences of delimiter
+                int idx = StringUtils.ordinalIndexOf(str, delimiter, count);
+                if (idx != StringUtils.INDEX_NOT_FOUND) {
+                    return StringData.fromString(str.substring(0, idx));
+                }
+            } else {
+                // get index of the last -count-th occurrences of delimiter
+                int idx = StringUtils.lastOrdinalIndexOf(str, delimiter, -count);
+                if (idx != StringUtils.INDEX_NOT_FOUND) {
+                    return StringData.fromString(str.substring(idx + delimiter.length()));
+                }
+            }
+        } catch (Throwable t) {
+            throw new FlinkRuntimeException(t);
+        }
+
+        // can not find enough delimiter
+        return expr;
+    }
+
+    public @Nullable byte[] eval(
+            @Nullable byte[] expr, @Nullable byte[] delim, @Nullable Integer count) {
+        if (expr == null || delim == null || count == null) {
+            return null;
+        }
+        if (expr.length == 0 || delim.length == 0 || count == 0) {
+            return new byte[] {};
+        }
+
+        if (count > 0) {
+            int idx = -1;
+            while (count > 0) {
+                // get index of the next occurrence of delim
+                idx = find(expr, idx + 1, delim);
+                if (idx >= 0) {
+                    --count;
+                } else {
+                    // can not find enough delim
+                    return expr;
+                }
+            }
+            if (idx == 0) {
```

Comment:
Can you help explain this logic to me? What case will will occur idx == 0, then return a empty byte array?

Reply:
`expr` starts with `delim` and `count = 1`. Here is an example:
```SQL
select substring_index('www.apache.org', 'w', 1);
 
```